### ISSUE TITLE
Stash not backing up all files

### ISSUE THREAD

**Issue by codejamninja** [stashed/stash#899]
Stash does not back up all of my files. Some of the files it backs up and some it doesn't. I can't figure out why. I change the user owner from `git` to `root` and it backed up more folders, but still didn't backup everything.

**Comment by codejamninja** [comment]
These are the files it is backing up . . .

```
/_data
/_data/data
/_data/data/shared
/_data/data/shared/pages
/_data/etc
/_data/etc/ssh_host_ecdsa_key.pub
/_data/etc/ssh_host_ed25519_key.pub
/_data/etc/ssh_host_rsa_key.pub
/_data/etc/trusted-certs
```

However, these are all the files contained in the volume

```
tree /_data

.
├── data
│   ├── builds
│   ├── shared
│   │   ├── artifacts
│   │   ├── cache
│   │   ├── dependency_proxy
│   │   ├── external-diffs
│   │   ├── lfs-objects
│   │   ├── packages
│   │   ├── pages
│   │   └── tmp
│   ├── ssh
│   │   └── authorized_keys
│   └── uploads
├── etc
│   ├── gitlab-secrets.json
│   ├── gitlab.rb
│   ├── ssh_host_ecdsa_key
│   ├── ssh_host_ecdsa_key.pub
│   ├── ssh_host_ed25519_key
│   ├── ssh_host_ed25519_key.pub
│   ├── ssh_host_rsa_key
│   ├── ssh_host_rsa_key.pub
│   └── trusted-certs
├── git
│   └── repositories
│       └── @hashed
│           └── 6b
│               └── 86
│                   ├── 6b86b273ff34fce19d6b804eff5a3f5747ada4eaa22f1d49c01e52ddb7875b4b.git
│                   │   ├── HEAD
│                   │   ├── branches
│                   │   ├── config
│                   │   ├── description
│                   │   ├── hooks
│                   │   │   ├── applypatch-msg.sample
│                   │   │   ├── commit-msg.sample
│                   │   │   ├── fsmonitor-watchman.sample
│                   │   │   ├── post-update.sample
│                   │   │   ├── pre-applypatch.sample
│                   │   │   ├── pre-commit.sample
│                   │   │   ├── pre-push.sample
│                   │   │   ├── pre-rebase.sample
│                   │   │   ├── pre-receive.sample
│                   │   │   ├── prepare-commit-msg.sample
│                   │   │   └── update.sample
│                   │   ├── info
│                   │   │   └── exclude
│                   │   ├── language-stats.cache
│                   │   ├── objects
│                   │   │   ├── 9d
│                   │   │   │   └── 0de6245c6d75c4d8e06c6a39c3f37628a1690c
│                   │   │   ├── b6
│                   │   │   │   └── fc4c620b67d95f953a5c1c1230aaab5db5a1b0
│                   │   │   ├── d2
│                   │   │   │   └── 5592c38ef63a211bf1d582f0d5e6c015438854
│                   │   │   ├── info
│                   │   │   └── pack
│                   │   └── refs
│                   │       ├── heads
│                   │       │   └── master
│                   │       ├── keep-around
│                   │       │   └── 9d0de6245c6d75c4d8e06c6a39c3f37628a1690c
│                   │       └── tags
│                   └── 6b86b273ff34fce19d6b804eff5a3f5747ada4eaa22f1d49c01e52ddb7875b4b.wiki.git
│                       ├── HEAD
│                       ├── branches
│                       ├── config
│                       ├── description
│                       ├── hooks
│                       │   ├── applypatch-msg.sample
│                       │   ├── commit-msg.sample
│                       │   ├── fsmonitor-watchman.sample
│                       │   ├── post-update.sample
│                       │   ├── pre-applypatch.sample
│                       │   ├── pre-commit.sample
│                       │   ├── pre-push.sample
│                       │   ├── pre-rebase.sample
│                       │   ├── pre-receive.sample
│                       │   ├── prepare-commit-msg.sample
│                       │   └── update.sample
│                       ├── info
│                       │   └── exclude
│                       ├── objects
│                       │   ├── info
│                       │   └── pack
│                       └── refs
│                           ├── heads
│                           └── tags
└── lost+found
``

**Comment by hossainemruz** [comment]
Can you please share the YAMLs of your workload and BackupConfiguration crd?

**Comment by codejamninja** [comment]
_backupconfiguration_
```sh
 kubectl describe backupconfiguration deployment-gitlab-gitlab-gitlab -n gitlab
```

```yaml
Name:         deployment-gitlab-gitlab-gitlab
Namespace:    gitlab
Labels:       <none>
Annotations:  <none>
API Version:  stash.appscode.com/v1beta1
Kind:         BackupConfiguration
Metadata:
  Creation Timestamp:  2019-09-17T06:54:57Z
  Finalizers:
    stash.appscode.com
  Generation:  1
  Owner References:
    API Version:           apps/v1
    Block Owner Deletion:  false
    Kind:                  Deployment
    Name:                  gitlab-gitlab-gitlab
    UID:                   0fc02fb8-d918-11e9-865a-06d69e0f0b3e
  Resource Version:        2660797
  Self Link:               /apis/stash.appscode.com/v1beta1/namespaces/gitlab/backupconfigurations/deployment-gitlab-gitlab-gitlab
  UID:                     0fd177fd-d918-11e9-865a-06d69e0f0b3e
Spec:
  Repository:
    Name:  deployment-gitlab-gitlab-gitlab
  Retention Policy:
    Keep Last:  5
    Name:       retention-policy
    Prune:      true
  Runtime Settings:
  Schedule:  0 0 * * *
  Target:
    Paths:
      /_data
    Ref:
      API Version:  apps/v1
      Kind:         Deployment
      Name:         gitlab-gitlab-gitlab
    Volume Mounts:
      Mount Path:  /_data
      Name:        data
  Task:
  Temp Dir:
Events:  <none>
```

_deployment_
```yaml
apiVersion: apps/v1beta2
kind: Deployment
metadata:
  annotations:
    deployment.kubernetes.io/revision: "2"
    field.cattle.io/publicEndpoints: '[{"addresses":["34.214.90.198"],"port":80,"protocol":"HTTP","serviceName":"gitlab:gitlab-gitlab-gitlab","ingressName":"gitlab:gitlab-gitlab-gitlab","hostname":"git.codejam.ninja","path":"/","allNodes":true}]'
    stash.appscode.com/backup-blueprint: gitlab-gitlab
    stash.appscode.com/last-applied-backupconfiguration: |
      {"kind":"BackupConfiguration","apiVersion":"stash.appscode.com/v1beta1","metadata":{"name":"deployment-gitlab-gitlab-gitlab","namespace":"gitlab","selfLink":"/apis/stash.appscode.com/v1beta1/namespaces/gitlab/backupconfigurations/deployment-gitlab-gitlab-gitlab","uid":"0fd177fd-d918-11e9-865a-06d69e0f0b3e","resourceVersion":"2660795","generation":1,"creationTimestamp":"2019-09-17T06:54:57Z","ownerReferences":[{"apiVersion":"apps/v1","kind":"Deployment","name":"gitlab-gitlab-gitlab","uid":"0fc02fb8-d918-11e9-865a-06d69e0f0b3e","blockOwnerDeletion":false}]},"spec":{"schedule":"0 0 * * *","repository":{"name":"deployment-gitlab-gitlab-gitlab"},"task":{},"target":{"ref":{"apiVersion":"apps/v1","kind":"Deployment","name":"gitlab-gitlab-gitlab"},"paths":["/_data"],"volumeMounts":[{"name":"data","mountPath":"/_data"}]},"retentionPolicy":{"name":"retention-policy","keepLast":5,"prune":true},"runtimeSettings":{},"tempDir":{}}}
    stash.appscode.com/target-paths: /_data
    stash.appscode.com/volume-mounts: data:/_data
  creationTimestamp: "2019-09-17T06:54:57Z"
  generation: 3
  labels:
    app: gitlab-gitlab
    chart: gitlab-0.3.0
    heritage: Tiller
    io.cattle.field/appId: gitlab
    release: gitlab
  name: gitlab-gitlab-gitlab
  namespace: gitlab
  resourceVersion: "2661234"
  selfLink: /apis/apps/v1beta2/namespaces/gitlab/deployments/gitlab-gitlab-gitlab
  uid: 0fc02fb8-d918-11e9-865a-06d69e0f0b3e
spec:
  progressDeadlineSeconds: 600
  replicas: 1
  revisionHistoryLimit: 2
  selector:
    matchLabels:
      app: gitlab-gitlab
      release: gitlab
  strategy:
    type: Recreate
  template:
    metadata:
      annotations:
        stash.appscode.com/last-applied-backupconfiguration-hash: "17930584425617363296"
      creationTimestamp: null
      labels:
        app: gitlab-gitlab
        release: gitlab
    spec:
      containers:
      - env:
        - name: GITLAB_OMNIBUS_CONFIG
          valueFrom:
            configMapKeyRef:
              key: omnibus_config
              name: gitlab-gitlab
        - name: EXTERNAL_URL
          valueFrom:
            configMapKeyRef:
              key: base_url
              name: gitlab-gitlab
        - name: DB_HOST
          valueFrom:
            configMapKeyRef:
              key: postgres_host
              name: gitlab-gitlab
        - name: DB_USER
          valueFrom:
            secretKeyRef:
              key: POSTGRES_USER
              name: gitlab-gitlab-postgres-auth
        - name: DB_DATABASE
          valueFrom:
            configMapKeyRef:
              key: postgres_database
              name: gitlab-gitlab
        - name: REDIS_HOST
          valueFrom:
            configMapKeyRef:
              key: redis_host
              name: gitlab-gitlab
        - name: GITLAB_ROOT_PASSWORD
          valueFrom:
            secretKeyRef:
              key: password
              name: gitlab-gitlab
        - name: DB_PASSWORD
          valueFrom:
            secretKeyRef:
              key: POSTGRES_PASSWORD
              name: gitlab-gitlab-postgres-auth
        image: gitlab/gitlab-ce:12.2.5-ce.0
        imagePullPolicy: IfNotPresent
        livenessProbe:
          failureThreshold: 6
          httpGet:
            path: /health
            port: container
            scheme: HTTP
          initialDelaySeconds: 120
          periodSeconds: 10
          successThreshold: 1
          timeoutSeconds: 5
        name: gitlab-gitlab-gitlab
        ports:
        - containerPort: 80
          name: container
          protocol: TCP
        readinessProbe:
          failureThreshold: 6
          httpGet:
            path: /readiness
            port: container
            scheme: HTTP
          initialDelaySeconds: 30
          periodSeconds: 10
          successThreshold: 1
          timeoutSeconds: 5
        resources:
          requests:
            cpu: 300m
            memory: 1288490188800m
        terminationMessagePath: /dev/termination-log
        terminationMessagePolicy: File
        volumeMounts:
        - mountPath: /_data
          name: data
        - mountPath: /gitlab-data
          name: data
          subPath: data
        - mountPath: /etc/gitlab
          name: data
          subPath: etc
        - mountPath: /var/opt/gitlab/git-data
          name: data
          subPath: git
      - args:
        - run-backup
        - --backup-configuration=deployment-gitlab-gitlab-gitlab
        - --secret-dir=/etc/stash/repository/secret
        - --enable-cache=true
        - --max-connections=0
        - --metrics-enabled=true
        - --pushgateway-url=http://stash.kube-system.svc:56789
        - --enable-status-subresource=true
        - --use-kubeapiserver-fqdn-for-aks=true
        - --enable-analytics=true
        - --logtostderr=true
        - --alsologtostderr=false
        - --v=3
        - --stderrthreshold=0
        env:
        - name: NODE_NAME
          valueFrom:
            fieldRef:
              apiVersion: v1
              fieldPath: spec.nodeName
        - name: POD_NAME
          valueFrom:
            fieldRef:
              apiVersion: v1
              fieldPath: metadata.name
        image: appscode/stash:v0.9.0-rc.0
        imagePullPolicy: IfNotPresent
        name: stash
        resources: {}
        terminationMessagePath: /dev/termination-log
        terminationMessagePolicy: File
        volumeMounts:
        - mountPath: /etc/stash
          name: stash-podinfo
        - mountPath: /etc/stash/repository/secret
          name: stash-secret-volume
        - mountPath: /tmp
          name: tmp-dir
        - mountPath: /_data
          name: data
      dnsPolicy: ClusterFirst
      restartPolicy: Always
      schedulerName: default-scheduler
      securityContext: {}
      terminationGracePeriodSeconds: 30
      volumes:
      - name: data
        persistentVolumeClaim:
          claimName: gitlab-gitlab
      - emptyDir: {}
        name: tmp-dir
      - downwardAPI:
          defaultMode: 420
          items:
          - fieldRef:
              apiVersion: v1
              fieldPath: metadata.labels
            path: labels
        name: stash-podinfo
      - name: stash-secret-volume
        secret:
          defaultMode: 420
          secretName: gitlab-gitlab-backup
status:
  availableReplicas: 1
  conditions:
  - lastTransitionTime: "2019-09-17T06:56:40Z"
    lastUpdateTime: "2019-09-17T06:56:40Z"
    message: Deployment has minimum availability.
    reason: MinimumReplicasAvailable
    status: "True"
    type: Available
  - lastTransitionTime: "2019-09-17T06:54:57Z"
    lastUpdateTime: "2019-09-17T06:56:40Z"
    message: ReplicaSet "gitlab-gitlab-gitlab-67558bb698" has successfully progressed.
    reason: NewReplicaSetAvailable
    status: "True"
    type: Progressing
  observedGeneration: 3
  readyReplicas: 1
  replicas: 1
  updatedReplicas: 1
```

**Comment by codejamninja** [comment]
It turns out this was a file permission issue. Any advice on how to proceed?

**Comment by hossainemruz** [comment]
For future reference, here is what happened and how to solve it.

**What happened:**

Stash runs backup sidecar as non-root user. Hence, if the targeted files are not readable to non-root user, Stash won't be able to backup them.

**How to solve it:**

You can run Stash sidecar as root user by setting `spec.runtimeSettings.container.securityContext` field in `BackupConfiguration`.

A sample `BackupConfiguration` is given below.
```yaml
apiVersion: stash.appscode.com/v1beta1
kind: BackupConfiguration
metadata:
  name: deployment-backup
  namespace: demo
spec:
  repository:
    name: gcs-repo
  schedule: "*/5 * * * *"
  target:
    ref:
      apiVersion: apps/v1
      kind: Deployment
      name: stash-demo
    volumeMounts:
    - name: source-data
      mountPath: /source/data
    paths:
    - /source/data
  runtimeSettings:
    container:
      securityContext:
        runAsUser: 0
        runAsGroup: 0
  retentionPolicy:
    name: 'keep-last-5'
    keepLast: 5
    prune: true
```

Also, please read this doc for additional information: https://github.com/stashed/docs/blob/master/docs/guides/latest/advanced-use-case/ownership.md

**Comment by peterholak** [comment]
Is there any way to automatically detect such a situation? Could stash perhaps expose it in the metrics (e.g. number of inaccessible files)?

**Comment by davidhiendl** [comment]
I just ran into this when verifying backups myself. I believe BackupSessions should not be marked as successful when any error regarding permissions is raised by Restic, it should at least say "Successful with Warnings". It is way to easy to miss this kind of error especially if only a fraction of the files are affected. The only reason I noticed it is that I compared the live data to a restored backup with diff.
